### PR TITLE
Ray TPU Webhook Autoscaling Changes

### PR DESCRIPTION
**Pull request by ryanaoleary**
Update Ray TPU Webhook to be compatible with Kuberay CRD changes (https://github.com/ray-project/kuberay/pull/1834, https://github.com/ray-project/kuberay/pull/1920, https://github.com/ray-project/kuberay/pull/1913) in order to support TPU pod autoscaling. This PR has been manually tested by deploying the webhook, creating a RayCluster with `replicas = 2` and `numOfHosts = 2` and verifying that 4 pods were scaled up and that the webhook correctly injects all labels, pod affinity constraints, and unique environment vars for each slice.

### CODE REVIEW

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 1]
```diff
@@ -107,57 +119,89 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
...
+
+	*patches = append(*patches, labelPatch)
+}
+
+// inject pod affinity and anti-affinity scheduling constraints using multiHostReplica label
+func injectPodAffinity(replicaIndex int, workerGroupName string, patches *[]patch) {
+	podAffinityPatch := patch{"op": "add"}
+	podAffinityPath := "/spec/affinity"
+	podAntiAffinityPatch := patch{"op": "add"}
+	podAntiAffinityPath := "/spec/affinity"
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:173]: /spec/antiaffinity?

File: applications/ray/TPU_guide.md [hunk 2]
```diff
@@ -54,7 +54,7 @@ The TPU Initialization Webhook automatically injects the `TPU_WORKER_ID`, `TPU_N
...
 
-To deploy a multi-host Ray Cluster, modify the `worker` spec [here](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/modules/kuberay-cluster/kuberay-tpu-values.yaml) by changing the `cloud.google.com/gke-tpu-topology` `nodeSelector` to a multi-host topology. Set the `replicas` field in the `worker` spec to the number of hosts specified by your chosen topology. For v4 TPUs, each TPU VM has access to 4 TPU chips. Therefore, you can calculate the number of TPU VM hosts by taking the product of the topology and dividing by 4 (i.e. a 2x2x4 TPU podslice will have 4 TPU VM hosts).
+To deploy a multi-host Ray Cluster, modify the `worker` spec [here](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/modules/kuberay-cluster/kuberay-tpu-values.yaml) by changing the `cloud.google.com/gke-tpu-topology` `nodeSelector` to a multi-host topology. Set the `numOfHosts` field in the `worker` spec to the number of hosts specified by your chosen topology. For v4 TPUs, each TPU VM has access to 4 TPU chips. Therefore, you can calculate the number of TPU VM hosts by taking the product of the topology and dividing by 4 (i.e. a 2x2x4 TPU podslice will have 4 TPU VM hosts).
```
Comment (richardsliu) [inline applications/ray/TPU_guide.md:57]: We should call out in this file that the webhook requires the v1.1 version of Kuberay, and will not work with the v1.0 version.

Also this requires GKE 1.28+.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 3]
```diff
@@ -251,85 +284,72 @@ func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissio
...
-	return admissionResponse, nil
+	return ""
 }
 
-func hasWorkerID(container corev1.Container) bool {
-	if container.Env != nil && len(container.Env) > 0 {
-		for _, envVar := range container.Env {
-			if envVar.Name == "TPU_WORKER_ID" {
-				return true
+// get next replica ID to assign a pod to
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:298]: This should be more specific - I think this is looking for the next lowest-index pod slice that has a missing created pod?

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 4]
```diff
@@ -251,85 +284,72 @@ func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissio
...
+func getReplicaIndex() int {
+	if sliceToWorkers == nil {
+		return 0
+	}
+	next_lowest_id := math.MaxInt32
+	for slice, workerList := range sliceToWorkers {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:304]: Looks like sliceToWorkers could contain slice information from multiple Ray clusters, would that cause any issues here?

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 5]
```diff
@@ -251,85 +284,72 @@ func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissio
...
+				replacePod = true
+				nextLowestID = worker.workerIndex
+			}
+		}
+		// reassign next lowest TPU_WORKER_ID if pod has been deleted
+		if replacePod == true {
+			for _, worker := range sliceToWorkers[podSlice] {
+				// set worker.isRunning to true now that pod is being re-created
+				if worker.workerIndex == nextLowestID {
+					worker.isRunning = true
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:341]: What if this pod creation failed?

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 1 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:173]: The path is `/spec/affinity/podAffinity` and `/spec/affinity/podAntiAffinity` for podAffinity and podAntiAffinity, I'll remove the duplicate var though and use `podAffinityPath` for both patches.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 4 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:303]: That's a good catch, I added in a check to make sure that it's only counting the pods/slices in the desired cluster.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 6]
```diff
@@ -20,10 +21,18 @@ import (
...
 	groupName      string
+	replicaIndex   int
+	numOfHosts     int32
+}
+
+// our representation of a worker pod
+type worker struct {
+	workerIndex  int  // TPU_WORKER_ID
+	replicaIndex int  // index of replica worker belongs to
+	isRunning    bool // true = pod is running, false = pod deleted or hasn't been created
```
[the remark this answers is not in the document]
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go]: Done.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 3 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go]: Done.

File: modules/kuberay-cluster/kuberay-tpu-values.yaml [hunk 7]
```diff
@@ -142,6 +142,7 @@ worker:
   replicas: 1
   minReplicas: 1 
   maxReplicas: 1
+  numOfHosts: 1
```
[the remark this answers is not in the document]
Comment (ryanaoleary) [reply modules/kuberay-cluster/kuberay-tpu-values.yaml:156]: I tested multiple hosts by just changing the numOfHosts field and then creating the cluster, I don't think there's an existing yaml template that by default sets numOfHosts > 1.

File: applications/ray/TPU_guide.md (hunk 2 above)
Comment (ryanaoleary) [reply applications/ray/TPU_guide.md:57]: Done.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 5 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go]: We sort of just assume that if it gets to this point the pod creation will succeed, the admission reponse is always set with `Allowed: true` and it's only possible for it to fail if the json.Marshal fails which should never happen.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 8]
```diff
@@ -80,12 +92,12 @@ func getNumTPUHosts(topology string) (int, error) {
 		chips *= dim
 	}
 	// number VMs = number chips / 4
-	return max(chips/4, 1), nil
+	return int32(max(chips/4, 1)), nil
```
[the remark this answers is not in the document]
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:95]: Based on: https://cloud.google.com/tpu/docs/supported-tpu-configurations#v5e_tpu_vm_type_comparison v5e TPU slices can contain 1, 4, or 8 chips per VM host. I added in a check for when `acceleratorType == "v5litepod` to determine # chips per host, but I'm looking into whether there's an easier way to do this.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 9]
```diff
@@ -79,13 +91,29 @@ func getNumTPUHosts(topology string) (int, error) {
...
+	// calculate the # of VMs using # of chips per host
+	acceleratorTypeValues := strings.Split(acceleratorType, "-")
+	var vms int32
+	if acceleratorTypeValues[0] == "v5litepod" {
+		// v5e TPU VMs can have 1, 4 or 8 chips
+		chipsPerHost, err := strconv.Atoi(acceleratorTypeValues[1])
+		if err != nil {
+			klog.Errorf("Unexpected acceleratorType: %s", acceleratorType)
+		}
+		if chipsPerHost > 8 {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:103]: Maybe just default it to 4?

**Review by richardsliu — APPROVED** [review]

**Comment by ryanaoleary** [comment]
> /gcbrun
Wait one sec before merging, I'm about to push a couple commits to resolve the last comments.

File: applications/ray/TPU_guide.md [hunk 10]
```diff
@@ -38,7 +38,7 @@ accelerator_type       = "nvidia-tesla-t4"
 
 ### Installing the TPU Initialization Webhook
 
-The TPU Initialization Webhook automatically injects the `TPU_WORKER_ID`, `TPU_NAME`, and `TPU_WORKER_HOSTNAMES` environment variables necessary for multi-host TPU clusters. The webhook needs to be installed once per GKE cluster. The instructions can be found [here](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/applications/ray/kuberay-tpu-webhook).
+The TPU Initialization Webhook automatically injects the `TPU_WORKER_ID`, `TPU_NAME`, and `TPU_WORKER_HOSTNAMES` environment variables necessary for multi-host TPU clusters. The webhook needs to be installed once per GKE cluster and requires a Kuberay Operator running v1.1 and GKE cluster version of 1.28+. The instructions can be found [here](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/applications/ray/kuberay-tpu-webhook).
```
[the remark this answers is not in the document]
Comment (ryanaoleary) [reply applications/ray/TPU_guide.md:41]: Done, added to the webhook README.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 9 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:103]: Done.

**Pull request merged**